### ISSUE TITLE
Code analysis ignores phpDoc type declarations for variables that are set as the first statement within a switch 'case' inside a function

### ISSUE THREAD

**Issue by tim-hk** [kalessil/phpinspectionsea#1243]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Extended) v3.0.11 |
| Language level | PHP 7.2  |

Current behaviour
--
I've come across a strange edge-case issue where IntelliJ IDEA seems to ignore phpDoc type-declarations for variables in a very specific circumstance:

* Where a variable is set as the first instruction of a switch 'case' statement
* AND where that switch is inside a function
* AND where the variable may be an object or null
* AND where an `@var` phpDoc type declaration has been supplied to indicate that the variable will be an instance of a class

In that very specific scenario, IDEA warns _"Null pointer exception may occur here."_ even though the `@var` declaration defines the type of the variable as non-null.  Adding any other code before the statement where the variable gets set (e.g. `$a = 1;`) resolves the problem.  A switch outside of a function does not have the same problem.

**Consider the following code:**

```php
<?php

/**
 * Class Foobar
 */
class Foobar {
    public function doFoo(): void {}
}

/**
 * @param bool $actually
 *
 * @return Foobar|null
 */
function getFoobar(bool $actually): ?Foobar
{
    // May return instance of Foobar or may return null
    return $actually ? new Foobar() : null;
}

test('bar');

/**
 * @param $foo
 */
function test($foo)
{
    switch ($foo) {
        case 'bar':
            /** @var Foobar $foobar */
            $foobar = getFoobar(true);
            $foobar->doFoo();
            break;
    }
}
```

In the above code, `$foobar` will always be an instance of the Foobar class, because it passes `true` to `getFoobar();`.  Because `getFoobar();` can return null in other cases (i.e. if you pass `false` to it), a null pointer exception warning occurs on the `$foobar->doFoo();` line.  However, the phpDoc comment `/** @var Foobar $foobar */` should suppress this, as it declares `$foobar` to be of type `Foobar`.

Adding any line of code to the case statement above `$foobar = getFoobar(true);` resolves the bug - it only affects variable declarations that are the first statement within a 'case', of a switch within a function.

Expected behaviour
--
Providing a phpDoc type declaration for a variable should always suppress inspection warnings that otherwise would have been shown due to an uncertain data type.

Environment details
---

IntelliJ IDEA 2018.3.5 (Ultimate Edition)
Build #IU-183.5912.21, built on February 26, 2019
Licensed to Timothy Evetts
Subscription is active until March 20, 2019
JRE: 1.8.0_152-release-1343-b28 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Windows 10 10.0

**Comment by orklah** [comment]
It is indeed a weird case. I tried your example, and a simple ";" before the phpdoc comment is enough to fix the issue...

Anyway, from what I see, there are two separate issues:
1. The resolver from PhpStorm is very good at adding potential types to a variable. But it's not that good at excluding types so when you add the PHPDoc with the type, it doesn't care and still see $foobar as "Foobar|null"(you see it when you hover the variable while pressing ctrl)
This can be fixed by placing the PHPDoc a line after the $foobar assignment. However, even with $foobar as "Foobar" only, the inspection still triggers.
2. When the type is only "Foobar" the inspection should not trigger anymore, plus there is this weird behaviour when the inspections triggers differently when the assignment is the first statement of the case block in a function. (If you remove the switch OR the function OR if you add a statement in the first line of the case block, the inspection is not triggered anymore)

**Comment by kalessil** [comment]
That seems to be a IDE bug, I created a ticket for this: https://youtrack.jetbrains.com/issue/WI-45867

**Comment by tim-hk** [comment]
I actually already raised this as an IDE bug, and it was knocked bag as being a PHP Inspections EA bug.  Here's my original issue here, which I raised before this one:

**Comment by kalessil** [comment]
Let see what will happen with my ticket as well =)
